Core/Src/processing.c

```diff
@@ -0,0 +1,365 @@
+/**
+ * @file processing.c
+ * @author Scott Abramson
+ * @brief Definitions for tasks for processing data.
+ * @version 0.1
+ * @date 2024-08-04
+ * 
+ * @copyright Copyright (c) 2024
+ * 
+ */
+
+#include "processing.h"
+#include "state_machine.h"
+#include "queues.h"
+#include "cerb_utils.h"
+#include "nero.h"
+#include "can_handler.h"
+#include "cerberus_conf.h"
+#include "dti.h"
+#include "queues.h"
+#include "serial_monitor.h"
+#include "bms.h"
+#include "emrax.h"
+#include "monitor.h"
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+#include <math.h>
+#include <stdlib.h>
+
+/* DO NOT ATTEMPT TO SEND TORQUE COMMANDS LOWER THAN THIS VALUE */
+#define MIN_COMMAND_FREQ  60 /* Hz */
+#define MAX_COMMAND_DELAY 1000 / MIN_COMMAND_FREQ /* ms */
+
+static float torque_limit_percentage = 1.0;
+
+#define TSMS_DEBOUNCE_PERIOD 500 /* ms */
+
+static bool tsms = false;
+osMutexId_t tsms_mutex;
+
+bool get_tsms()
+{
+	bool temp;
+	osMutexAcquire(tsms_mutex, osWaitForever);
+	temp = tsms;
+	osMutexRelease(tsms_mutex);
+	return temp;
+}
+
+void tsms_debounce_cb(void *arg)
+{
+	/* Set TSMS state to new debounced value */
+	osMutexAcquire(tsms_mutex, osWaitForever);
+	tsms = *((bool *)arg);
+	osMutexRelease(tsms_mutex);
+	/* Tell NERO allaboutit */
+	send_nero_msg();
+}
+
+osThreadId_t process_tsms_thread_id;
+const osThreadAttr_t process_tsms_attributes = {
+	.name = "ProcessTSMS",
+	.stack_size = 32 * 8,
+	.priority = (osPriority_t)osPriorityHigh,
+};
+
+void vProcessTSMS(void *pv_params)
+{
+	bool tsms_reading;
+	tsms_mutex = osMutexNew(NULL);
+
+	/* When the callback is called, the TSMS state will be set to the new reading in tsms_reading */
+	osTimerId_t tsms_debounce_timer =
+		osTimerNew(&tsms_debounce_cb, osTimerOnce, &tsms_reading, NULL);
+
+	for (;;) {
+		osThreadFlagsWait(TSMS_UPDATE_FLAG, osFlagsWaitAny,
+				  osWaitForever);
+
+		/* Get raw TSMS reading */
+		tsms_reading = get_tsms_reading();
```

Comment:
Ik this is what has to happen when you yeat threads but the amount of mutexing to propogate one boolean really worries me that we are going to delay/lock up the system do to so much mutexing that it will outweight the benefits of removing a thread.

Reply:
It wasn't a problem. I got rid of this anyway and just combined the TSMS processing and data collection because both tasks were small and quick.